### PR TITLE
From chapter 4 to 6

### PR DESCRIPTION
**Pull request by Vretamoza**

### CODE REVIEW

File: kv/lib/kv.ex [hunk 1]
```diff
@@ -1,18 +1,8 @@
...
-
-      iex> KV.hello()
-      :world
-
-  """
-  def hello do
-    :world
+  @impl true
+  def start(_type, _args) do
+    KV.Supervisor.start_link(name: KV.Supervisor) #Calls the function defined in Supervisor.ex
```
Comment (juanhurtado10) [inline kv/lib/kv.ex:6]: @Vretamoza what does a supervisor do?

File: kv/lib/kv/registry.ex [hunk 2]
```diff
@@ -6,50 +6,48 @@ defmodule KV.Registry do
   ## Defining GenServer Callbacks
 
   @impl true
-  def init(:ok) do
-    names = %{}
+  def init(table) do
+    # 3. We have replaced the names map by ets tables
+    names = :ets.new(table, [:named_table, read_concurrency: true])
```
Comment (juanhurtado10) [inline kv/lib/kv/registry.ex:11]: @Vretamoza could you please explain `ets`?

File: kv/lib/kv/registry.ex [hunk 3]
```diff
@@ -6,50 +6,48 @@ defmodule KV.Registry do
...
-      ref = Process.monitor(bucket)
-      refs = Map.put(refs, ref, name)
-      names = Map.put(names, name, bucket)
-      {:noreply, {names, refs}}
+  def handle_call({:create, name}, _from, {names, refs}) do
+    case lookup(names, name) do
+      {:ok, pid} ->
+        {:reply, pid, {names, refs}}
+      :error ->
+        {:ok, pid} = DynamicSupervisor.start_child(KV.BucketSupervisor, KV.Bucket)
```
Comment (juanhurtado10) [inline kv/lib/kv/registry.ex:22]: @Vretamoza what is the difference between a Supervisor and a DynamicSupervisor?

File: kv/lib/kv/supervisor.ex [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+  end
+
+  @impl true
+  def init(:ok) do
+    children = [
+      {DynamicSupervisor, name: KV.BucketSupervisor, strategy: :one_for_one},
+      {KV.Registry, name: KV.Registry}
+  ]
+
+    Supervisor.init(children, strategy: :one_for_all)
```
Comment (juanhurtado10) [inline kv/lib/kv/supervisor.ex:15]: @Vretamoza what does the `strategy: :one_for_all` mean?

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]
@Vretamoza Well job done, some comments were added 🚀 !!

File: kv/lib/kv.ex (hunk 1 above)
Comment (Vretamoza) [reply kv/lib/kv.ex:6]: A supervisor is a process that is responsable of supervising other processes. It starts the processes, restarts them if one of them crashes or finishes, in order to keep them available, and also shuts them out when necessary.

File: kv/lib/kv/registry.ex (hunk 2 above)
Comment (Vretamoza) [reply kv/lib/kv/registry.ex:11]: ETS is the Erlang Term Storage, this module provides a large data storage capacity, which is stored in tables that can be quicly accessed. ETS tables are owned by the processes that create them and according to certain parameters, they can be read, modified, or not, by other processes.

File: kv/lib/kv/registry.ex (hunk 3 above)
Comment (Vretamoza) [reply kv/lib/kv/registry.ex:22]: The difference is that Supervisors know all the children processes they are going to supervise previously, a DynamicSupervisor monitors processes that are created dynamically, it doesn't expect a list of children, instead, offers functions that allows us to create children processes.

File: kv/lib/kv/supervisor.ex (hunk 4 above)
Comment (Vretamoza) [reply kv/lib/kv/supervisor.ex:15]: It means that when a child process terminates, all other child processes are also terminated and then they are all restarted.

**Pull request merged**